### ISSUE TITLE
Improve logging when dual ratio test fails

### ISSUE THREAD

**Issue by NPC** [ERGO-Code/HiGHS#1484]
Hello,

I've encountered this issue that I can't find an explanation for, nor can I identify the cause by looking at the log, can you please help?

Here's the LP MPS file, running on Windows 11 from command line (same result called from the DLL via C# interface): [LP-problem-2023-11-03.zip](https://github.com/ERGO-Code/HiGHS/files/13254491/LP-problem-2023-11-03.zip)

Run via `highs.exe --parallel on LP-problem.mps`, output:  
```
Running HiGHS 1.6.0: Copyright (c) 2023 HiGHS under MIT licence terms
LP   LP-problem has 18247 rows; 57099 cols; 107808 nonzeros
Presolving model
9111 rows, 30394 cols, 53144 nonzeros
8939 rows, 29003 cols, 50965 nonzeros
Presolve : Reductions: rows 8939(-9308); columns 29003(-28096); elements 50965(-56843)
Solving the presolved LP
Using EKK parallel dual simplex solver - PAMI with concurrency of 8
  Iteration        Objective     Infeasibilities num(sum)
          0    -6.3260410194e+02 Ph1: 1(1); Du: 1(632.604) 0s
Model   status      : Not Set
HiGHS run time      :          0.14
```

I tried parallel off, presolve off, versions 1.5.3 and 1.4.0, no difference.

As I understand, “Not Set” means that the algorithm ended unexpectedly, before it could report on any errors or infeasibilities, is this correct? Could the values `Ph1: 1(1); Du: 1(632.604)` in “Infeasibilities” column provide some valuable insight?

Simpler problems continue to solve without issues, as well as complex but different problems, these's something off about this one.

I'll be grateful for any help.

**Comment by jajhall** [comment]
How odd. I've reproduced this behaviour with the latest (dev) version of HiGHS

**Comment by jajhall** [comment]
Your problem is badly scaled. You have very large costs 

> 5055 values satisfy 10^( 11) <= v < 10^( 12)

upper bounds on variables

> 12 values satisfy 10^( 10) <= v < 10^( 11)

and RHS values for equality constraints

> 4 values satisfy 10^( 10) <= v < 10^( 11)

The HiGHS dual simplex solver fails due to the excessive costs. I see that you're running HiGHS from the command line, so the logging output that you quote is all that you're told. If calling HiGHS from an application, you'd get an error code returned from Highs::run()

With the option `log_dev_level=1`, set in a file (say) `Options.set`, and running

> bin/highs --options_file=Options.set LP-problem-2023-11-03.mps

HiGHS gives the following logging information

> Cost perturbation for LP-problem-2023-11-03
   Initially have 29003 nonzero costs (100%) with min / average / max = 26 / 6.23951e+12 / 2.00573e+14
   Large so set max_abs_cost = sqrt(sqrt(max_abs_cost)) = 3763.29
   Perturbation column base = 0.00188165
   Perturbation row    base = 1e-12
       Iteration        Objective     Infeasibilities num(sum)
DuPh1          0    -6.3260410194e+02 Ph1: 1(1); Du: 1(632.604) No reason
DuPh1          1     0.0000000000e+00 Ph1: 0(0) Possibly optimal
DuPh2          1     2.6075283087e+22 Pr: 4928(2.86593e+11) No reason
DuPh2        267     2.6236821231e+22 Pr: 4740(6.43333e+10) Synthetic clock
DuPh2        526     2.6762934176e+22 Pr: 4580(4.59647e+10) Synthetic clock
DualChuzC:     No change in loop 2 so return error
DualChuzC:     workCount = 43; selectTheta=4.60303e+13; remainTheta=4.60303e+13
DualChuzC:     workDataNorm = 7.39874; workDualNorm = 3.70191e+15
DuPh2        669     2.7326666760e+22 Pr: 4538(4.17281e+10) Choose column failure
DualChuzC:     No change in loop 2 so return error
DualChuzC:     workCount = 43; selectTheta=4.60303e+13; remainTheta=4.60303e+13
DualChuzC:     workDataNorm = 7.39874; workDualNorm = 3.70191e+15
dual-phase-2-not-solved
HEkkDual::solve return of HighsStatus::Error
Simplex iterations: DuPh1 1; DuPh2 668; Total 669
solveLpSimplex return of HighsStatus::Error
callSolveLp return of HighsStatus::Error
Model   status      : Not Set
HiGHS run time      :          0.11

This is after your constraint matrix has been scaled - which is necessary as (originally) there is a very large range of values

> Matrix sparsity of dimension 107808 with 107808 nonzeros (100%) in [   1.77e-09,           1]
       73763 values satisfy 10^(  0) <= v < 10^(  1)
       33123 values satisfy 10^( -4) <= v < 10^( -3)
          21 values satisfy 10^( -6) <= v < 10^( -5)
         103 values satisfy 10^( -7) <= v < 10^( -6)
         798 values satisfy 10^( -8) <= v < 10^( -7)

Unfortunately, scaling up the small values increases some of the costs used by the solver by a couple of orders of magnitude. This means that, using 64 bit doubles, there's not enough numerical precision for the dual simplex algorithm to run correctly. Hence the failure.

**Comment by jajhall** [comment]
Essentially what's happening is that you're asking HiGHS to get dual feasibility to within a tolerance of `1e-7` using values of order `1e14`, a relative difference that's `1e5` greater than machine precision of `1e-16`. 

To you I suggest scaling your costs down by `1e6`

To me, I suggest a more informative logging message when this error occurs

**Comment by NPC** [comment]
@jajhall Thank you, this is very helpful! I'll look into cost scaling, this is something we haven't considered (or run into issues with) before. I actually call HiGHS from C#, but it didn't return an error other than kNotset.

**Comment by jajhall** [comment]
I'm not a C# user, and the C# interface was written externally. However (in `src/interfaces/highs_csharp_api.cs`) I see

>     public HighsStatus run()
    {
        return (HighsStatus)HighsLpSolver.Highs_run(this.highs);
    }

which is calling the C API method 

> HighsInt Highs_run(void* highs) { return (HighsInt)((Highs*)highs)->run(); }

which is calling the native C++ Highs class method

> HighsStatus Highs::run()

So, the status `HighsStatus::kError` returned from C++ should get passed down (eventually!) to C#. Do you see this value?


so I'd expect the error status returned by

**Comment by jajhall** [comment]
Closed by #1485

**Comment by EamonHetherton** [comment]
This turned up while tracking down a MIP that reports a suboptimal solution as Optimal at a zero gap — the ratio-test failure here is the trigger, though the defect I'm reporting is in how the MIP handles it (#3180), not in the behaviour described here. Two things that might still be useful on this thread: a sharper statement of what stalls, and a much smaller reproducer.

### What actually stalls

The dual simplex EXPAND ratio test stalls and returns `Solve error` on an LP that HiGHS solves to
optimality with presolve off, with the primal simplex, or with IPM. The stall happens once reduced
costs are large enough that adding the dual tolerance to them is a no-op in floating point.

[chuzc-1.mps.txt](https://github.com/user-attachments/files/30441343/chuzc-1.mps.txt) is 14 rows x 28 columns, 160 nonzeros, and reproduces under **default options**:

```
$ highs chuzc-1.mps
ERROR:   Dual simplex ratio test failed due to excessive dual values:
         consider scaling down the LP objective coefficients
Model status        : Solve error

$ highs --presolve off chuzc-1.mps
Model status        : Optimal        Objective value : -1.7568155358e+13
```

The failure is seed dependent: it occurs on `random_seed = 0`, which is the default, and not on
seeds 1, 2 or 3.

### Cause

`HEkkDualRow::chooseFinalWorkGroupQuad()`:

```cpp
while (selectTheta < kMaxSelectTheta) {
  double remainTheta = kInitialRemainTheta;
  for (HighsInt i = workCount; i < fullCount; i++) {
    ...
    if (dual <= selectTheta * value) {              // tight satisfy
      swap(workData[workCount++], workData[i]);
      totalChange += value * (workRange[iCol]);
    } else if (dual + Td < remainTheta * value) {
      remainTheta = (dual + Td) / value;
    }
  }
  workGroup.push_back(workCount);
  selectTheta = remainTheta;
  if ((workCount == prev_workCount) && (prev_selectTheta == selectTheta) &&
      (prev_remainTheta == remainTheta)) {
    ...
    return false;                                    // no progress
  }
  ...
}
```

On one pass a column sets `selectTheta = (dual + Td) / value`. On the next pass that column is
expected to satisfy `dual <= selectTheta * value`, because `selectTheta * value` is `dual + Td` and
`Td > 0`.

That reasoning fails when `Td` is negligible against `dual`. With the default dual feasibility
tolerance `Td = 1e-7`, once `|dual|` exceeds roughly `Td / eps = 1e-7 / 2.2e-16 ~ 4.5e8`, the sum
`dual + Td` rounds back to `dual` exactly. `selectTheta` becomes `dual / value`, and
`selectTheta * value` can then round to just below `dual`. The column is neither tight-satisfied nor
able to lower `remainTheta`; `workCount` and both thetas are unchanged; and the loop exits through
the no-progress guard, which `HEkkDual` turns into `kSolvePhaseError`.

Instrumenting a larger model that hits the same state (`log_dev_level = 2`):

```
DualChuzC:     No change in loop 2 so return error
DualChuzC:     workCount = 14; selectTheta=4.2494e+06; remainTheta=4.2494e+06
DualChuzC:     workDataNorm = 7496.33; workDualNorm = 5.87705e+10
```

`selectTheta` and `remainTheta` are equal, and `workDualNorm` is `5.9e10`, so `Td / dual` is about
`1.7e-18`, two orders of magnitude below machine epsilon. The attached model has an objective
coefficient range of `1.08e-06` to `2.61e+12` and reaches the same state.

The guard itself is doing its job — it prevents an infinite loop. The question is whether
`kSolveError` is the right outcome, given that the same LP is solved by three other configurations:

| how the LP is solved | result |
|---|---|
| dual simplex, presolve on — the default | `Solve error` |
| dual simplex, presolve off | `Optimal` |
| primal simplex | `Optimal` |
| IPM | `Optimal` |

### How it was found

By construction from the code rather than by search. Once the condition above was identified, LPs
were generated with objective coefficients large enough to make `dual + Td == dual`; between 5 and
24 of every 120 such LPs reproduce the failure. `chuzc-1.mps` is one of them reduced to 14 rows.
Generating without that insight produced nothing.

### Version

`latest` @ `bc30ccf1a1` (1.15.1). Also reproduces on `master` @ `04024d701f`.

### Why I am not filing this as a new issue

This issue already covers it: the diagnosis given here — dual feasibility to `1e-7` against values of
order `1e14`, a relative difference far beyond machine precision — is this same effect, and the
resolution was to warn rather than to change the ratio test. HiGHS duly warns
`Problem has some excessively large costs` on the attached model.

The reason it still matters is that this failure is reachable inside the MIP solver, where it is not
visible: a failed repair LP is treated as a bad solution and silently discarded, which can lose the
optimum and report a worse incumbent as `Optimal` at a zero gap. That is #3180, filed separately,
since the treatment of the failure is a defect independent of whether the failure itself is
expected.
